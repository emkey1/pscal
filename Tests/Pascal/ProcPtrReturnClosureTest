program ProcPtrReturnClosureTest;

type
  TGenerator = function: integer;

function MakeGenerator(start, step: integer): TGenerator;
var
  current: integer;

  function Next: integer;
  begin
    current := current + step;
    Next := current;
  end;
begin
  current := start;
  MakeGenerator := @Next;
end;

var
  genA, genB: TGenerator;
begin
  genA := MakeGenerator(0, 1);
  genB := MakeGenerator(10, -2);

  writeln('A->', genA());
  writeln('A->', genA());
  writeln('B->', genB());
  writeln('A->', genA());
  writeln('B->', genB());
end.
