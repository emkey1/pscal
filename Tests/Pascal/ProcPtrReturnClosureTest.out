A->1
A->2
B->8
A->3
B->6
